### PR TITLE
Convert Car.js script to module

### PR DESCRIPTION
**Pull request by looeee**
* renamed `Car.js` to `CarControls.js` and moved it to the **js/controls** folder. 
* updated `modularize.js`
* converted `CarControls.js` to module
* added `CarControls.d.ts`

This is the first time I've created a `.d.ts` file so if someone can check that over I'd appreciate it. 

There are a number of functions that are intended to be private but are exposed, such as `onKeyDown`, `onKeyUp` etc. I've left these out of the .ts file. Is that the correct approach? Otherwise I can refactor the script to make them user inaccessible.

### CODE REVIEW

File: examples/jsm/controls/CarControls.d.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import {
+  Object3D,
+} from '../../../src/Three';
+
+export class CarControls {
+  constructor( maxSpeed: number, acceleration: number, brakePower: number, turningRadius: number, keys: object );
```
Comment (Mugen87) [inline examples/jsm/controls/CarControls.d.ts]: All ctor parameters are optional so you can write them like so `maxSpeed?: number`.

**Comment by Mugen87** [comment]
You should also remove the old version of [Car.js](https://github.com/looeee/three.js/blob/convert_car_script_to_module/examples/js/Car.js) and update the respective example (so it uses `CarControls.js`).

http://raw.githack.com/looeee/three.js/convert_car_script_to_module/examples/webgl_materials_cars.html

**Comment by Mugen87** [comment]
> There are a number of functions that are intended to be private but are exposed, such as onKeyDown, onKeyUp etc. I've left these out of the .ts file. Is that the correct approach?

You have several options in this case and all are okay for me.

- Do not change the code but document everything.
- Mark private methods with an underscore. So `onKeyDown` becomes `_onKeyDown`.
- Completely remove them from the car prototype and turn them into functions placed inside the car closure.

**Comment by looeee** [comment]
> Or just leave it as it is and change the code in a different PR^^.

Let's do that one 😅 

> You should also remove the old version of Car.js and update the respective example (so it uses CarControls.js).

Oops, yes. Done.

File: examples/jsm/controls/CarControls.d.ts (hunk 1 above)
Comment (looeee) [reply examples/jsm/controls/CarControls.d.ts]: OK, done.

**Review by Mugen87 — APPROVED** [review]

**Comment by Mugen87** [comment]
Let's wait if @mrdoob is okay with moving the file into `js/controls`. I personally think it's fine like that.

**Comment by looeee** [comment]
Yeah, it's the place that makes sense to me. It takes user input and moves an object around, same as `TransformControls` and `DragControls`.

**Comment by WestLangley** [comment]
CarControls? That is application-specific, unlike the other controls. It is not a generic controller.

It should be embedded within the source code of the example that uses it, IMO.

**Comment by Mugen87** [comment]
Application specific code can also be placed in modules. If you think putting the file into `js/controls` is not good, I suggest to use a new directory `js/misc`.

**Comment by looeee** [comment]
Putting it in **misc/** is fine by me. Removing it altogether is also fine by me, [as I've said before](https://github.com/mrdoob/three.js/pull/13337#issuecomment-366405493).

However this is one of the prettier and more fun examples, and a good showcase of what's possible with three.js, which is a strong argument for keeping it.

I don't think we should embed a >300 line standalone function directly in the HTML though. The examples are supposed to be educational after all, and that's just bad practice.

**Comment by Mugen87** [comment]
Yes. `Car.js` is a nicely encapsulated component. Putting it directly into the HTML files is no good approach.

**Comment by WestLangley** [comment]
I like your example, too!

And yes, it is really a showcase. I think the examples should demonstrate specific three.js features. Showcases can be pretty, but users complain they are too complicated -- or at least they used to complain so.

Maybe `Car.js` can go in misc.

**Comment by Mugen87** [comment]
Then let's place it in `js/misc` 👍

**Comment by looeee** [comment]
OK, sounds good to me. I'd rather keep the name `CarControls.js` though since that most accurately describes what the module is for. `Car.js` used to be somewhat accurate before I removed the dependency on `BinaryLoader` and adapted the function to work with any generic car model, but it no longer is. Are you guys OK with that?

> I think the examples should demonstrate specific three.js features. 

In general, I agree. But I think it's fine to have some showcases that are a little bit more complex too. The examples for me have two equally important functions: 

1. education and reference for how to use three.js features
2. showcases for attracting new users

We do have the showcases on the threejs.org front page too of course, but they are generally commercial and have minified source code. Nice to look at, but you can't play around with them. More complex examples like this one, or the Ocean, Sky, Minecraft, Cloth and others help to bridge that gap.

**Comment by Mugen87** [comment]
> Are you guys OK with that?

👍

**Comment by looeee** [comment]
> Then let's place it in js/misc

OK, done!

**Comment by Mugen87** [comment]
Can you please resolve the merge conflict in `modularize.js`? 😇

**Comment by looeee** [comment]
> Can you please resolve the merge conflict in modularize.js?

Done done

**Comment by mrdoob** [comment]
Thanks guys! 😊

**Comment by Mugen87** [comment]
@looeee For some reason there seems to miss `Car.d.ts` in `jsm/misc`.

**Comment by Mugen87** [comment]
Added 😇

**Pull request merged**